File: app/src/main/java/com/komica/reader/adapter/ThreadAdapter.java
package com.komica.reader.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.komica.reader.R;
import com.komica.reader.model.Thread;
import androidx.recyclerview.widget.DiffUtil;
import java.util.ArrayList;
import java.util.List;

public class ThreadAdapter extends RecyclerView.Adapter<ThreadAdapter.ThreadViewHolder> {

    private List<Thread> threads;
    private OnThreadClickListener onThreadClickListener;

    public interface OnThreadClickListener {
        void onThreadClick(Thread thread);
        void onShareClick(Thread thread);
    }

    public ThreadAdapter(List<Thread> threads, OnThreadClickListener listener) {
        this.threads = new ArrayList<>(threads);
        this.onThreadClickListener = listener;
    }

    public void updateThreads(List<Thread> newThreads) {
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffUtil.Callback() {
            @Override
            public int getOldListSize() {
                return threads.size();
            }

            @Override
            public int getNewListSize() {
                return newThreads.size();
            }

            @Override
            public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                return threads.get(oldItemPosition).getUrl().equals(newThreads.get(newItemPosition).getUrl());
            }

            @Override
            public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                Thread oldItem = threads.get(oldItemPosition);
                Thread newItem = newThreads.get(newItemPosition);
                return oldItem.getPostNumber() == newItem.getPostNumber() &&
                       oldItem.getTitle().equals(newItem.getTitle()) &&
                       oldItem.getReplyCount() == newItem.getReplyCount() &&
                       (oldItem.getLastReplyTime() == null ? "" : oldItem.getLastReplyTime())
                           .equals(newItem.getLastReplyTime() == null ? "" : newItem.getLastReplyTime());
            }
        });

        this.threads.clear();
        this.threads.addAll(newThreads);
        diffResult.dispatchUpdatesTo(this);
    }

    @NonNull
    @Override
    public ThreadViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_thread, parent, false);
        return new ThreadViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ThreadViewHolder holder, int position) {
        Thread thread = threads.get(position);
        holder.bind(thread);
    }

    @Override
    public int getItemCount() {
        return threads.size();
    }

    class ThreadViewHolder extends RecyclerView.ViewHolder {
        private TextView threadPostNumber;
        private TextView threadTitle;
        private TextView threadAuthor;
        private TextView contentPreview;
        private TextView lastReplyTime;
        private TextView replyCount;
        private View shareButton;

        public ThreadViewHolder(@NonNull View itemView) {
            super(itemView);
            threadPostNumber = itemView.findViewById(R.id.threadPostNumber);
            threadTitle = itemView.findViewById(R.id.threadTitle);
            threadAuthor = itemView.findViewById(R.id.threadAuthor);
            contentPreview = itemView.findViewById(R.id.contentPreview);
            lastReplyTime = itemView.findViewById(R.id.lastReplyTime);
            replyCount = itemView.findViewById(R.id.replyCount);
            shareButton = itemView.findViewById(R.id.shareButton);
        }

        public void bind(Thread thread) {
            threadPostNumber.setText("No. " + thread.getPostNumber());
            threadTitle.setText(thread.getTitle());
            threadAuthor.setText(thread.getAuthor());
            replyCount.setText("回覆: " + thread.getReplyCount());

            if (thread.getContentPreview() != null && !thread.getContentPreview().isEmpty()) {
                contentPreview.setText(thread.getContentPreview());
                contentPreview.setVisibility(View.VISIBLE);
            } else {
                contentPreview.setVisibility(View.GONE);
            }
 
            if (thread.getLastReplyTime() != null && !thread.getLastReplyTime().isEmpty()) {
                lastReplyTime.setText(thread.getLastReplyTime());
                lastReplyTime.setVisibility(View.VISIBLE);
            } else {
                lastReplyTime.setVisibility(View.GONE);
            }

            itemView.setOnClickListener(v -> {
                if (onThreadClickListener != null) {
                    onThreadClickListener.onThreadClick(thread);
                }
            });
            
            shareButton.setOnClickListener(v -> {
                if (onThreadClickListener != null) {
                    onThreadClickListener.onShareClick(thread);
                }
            });
        }
    }
}

File: app/src/main/java/com/komica/reader/adapter/ThreadAdapter.kt
package com.komica.reader.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.komica.reader.R
import com.komica.reader.model.Thread

class ThreadAdapter(
    threads: List<Thread>,
    private val onThreadClickListener: OnThreadClickListener?
) : RecyclerView.Adapter<ThreadAdapter.ThreadViewHolder>() {

    private val threads: MutableList<Thread> = ArrayList(threads)

    interface OnThreadClickListener {
        fun onThreadClick(thread: Thread)
        fun onShareClick(thread: Thread)
    }

    fun updateThreads(newThreads: List<Thread>) {
        val diffResult = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize(): Int = threads.size

            override fun getNewListSize(): Int = newThreads.size

            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return threads[oldItemPosition].url == newThreads[newItemPosition].url
            }

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                val oldItem = threads[oldItemPosition]
                val newItem = newThreads[newItemPosition]
                return oldItem.postNumber == newItem.postNumber &&
                    oldItem.title == newItem.title &&
                    oldItem.replyCount == newItem.replyCount &&
                    oldItem.lastReplyTime == newItem.lastReplyTime
            }
        })

        threads.clear()
        threads.addAll(newThreads)
        diffResult.dispatchUpdatesTo(this)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ThreadViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_thread, parent, false)
        return ThreadViewHolder(view)
    }

    override fun onBindViewHolder(holder: ThreadViewHolder, position: Int) {
        holder.bind(threads[position])
    }

    override fun getItemCount(): Int = threads.size

    inner class ThreadViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val threadPostNumber: TextView = itemView.findViewById(R.id.threadPostNumber)
        private val threadTitle: TextView = itemView.findViewById(R.id.threadTitle)
        private val threadAuthor: TextView = itemView.findViewById(R.id.threadAuthor)
        private val contentPreview: TextView = itemView.findViewById(R.id.contentPreview)
        private val lastReplyTime: TextView = itemView.findViewById(R.id.lastReplyTime)
        private val replyCount: TextView = itemView.findViewById(R.id.replyCount)
        private val shareButton: View = itemView.findViewById(R.id.shareButton)

        private val titleTextSize: Float

        init {
            val prefs = itemView.context.getSharedPreferences("KomicaReader", android.content.Context.MODE_PRIVATE)
            titleTextSize = prefs.getFloat("theme_font_size", 16f)
        }

        fun bind(thread: Thread) {
            threadTitle.setTextSize(android.util.TypedValue.COMPLEX_UNIT_SP, titleTextSize)

            threadPostNumber.text = "No. ${thread.postNumber}"
            threadTitle.text = thread.title
            threadAuthor.text = thread.author
            replyCount.text = "回覆: ${thread.replyCount}"

            if (thread.contentPreview.isNotBlank()) {
                contentPreview.text = thread.contentPreview
                contentPreview.visibility = View.VISIBLE
            } else {
                contentPreview.visibility = View.GONE
            }

            if (thread.lastReplyTime.isNotBlank()) {
                lastReplyTime.text = thread.lastReplyTime
                lastReplyTime.visibility = View.VISIBLE
            } else {
                lastReplyTime.visibility = View.GONE
            }

            itemView.setOnClickListener {
                onThreadClickListener?.onThreadClick(thread)
            }

            shareButton.setOnClickListener {
                onThreadClickListener?.onShareClick(thread)
            }
        }
    }
}
